lalrpop/src/normalize/token_check/mod.rs

```diff
@@ -20,29 +20,38 @@ use collections::{map, Map};
 mod test;
 
 pub fn validate(mut grammar: Grammar) -> NormResult<Grammar> {
-    let (has_enum_token, all_literals) = {
+    let (has_enum_token, all_literals, match_to_user_name_map) = {
         let opt_match_token = grammar.match_token();
 
-        let (match_mappings, match_catch_all) = if let Some(mt) = opt_match_token {
-            let mut ret = map();
+        let (match_to_user_name_map, user_name_to_match_map, match_catch_all) = if let Some(mt) = opt_match_token {
```

Comment:
I still think this would be cleaner with one `Option<Struct>` then several individual variables.